### PR TITLE
chore: release v1.20.0

### PR DESCRIPTION
**Pull request by fedellen**
This merge PR implements delegated payments into the Turbo SDK and CLI

here are CLI examples, share Credits to a valid SOL/ETH/POL/AR/KYVE address:

```sh
node lib/esm/cli/cli.js share-credits --address 2cor8AZxm38uc9sVgGaXk4BU2EtB18FqwEq7ivwa7VUa --value 0.008 --wallet-file ../wallets/zarweave-keyfile-jaxl_dxqJ00gEgQazGASFXVRvO4h-Q0_vnaLtuOUoWU_dereks.json
{
  "message": "Created credit share approval!",
  "approvalDataItemId": "pdXRmDue07mX77m7GcgAz2z0esym8H_2-hshs8H454k",
  "approvedAddress": "2cor8AZxm38uc9sVgGaXk4BU2EtB18FqwEq7ivwa7VUa",
  "approvedWincAmount": "8000000000",
  "creationDate": "2024-10-31T15:04:15.898Z",
  "payingAddress": "jaxl_dxqJ00gEgQazGASFXVRvO4h-Q0_vnaLtuOUoWU",
  "usedWincAmount": "0"
}
```

controlled winc is reflected in balance command:

```sh
> node lib/esm/cli/cli.js balance --address jaxl_dxqJ00gEgQazGASFXVRvO4h-Q0_vnaLtuOUoWU
Turbo Balance for Native Address "jaxl_dxqJ00gEgQazGASFXVRvO4h-Q0_vnaLtuOUoWU"
Effective Credits: 1.9325555055
Credits Shared to Other Wallets: 0.016
```

list all given or received approvals:

```sh
> node lib/esm/cli/cli.js list-approvals --address jaxl_dxqJ00gEgQazGASFXVRvO4h-Q0_vnaLtuOUoWU
{
  "message": "Credit share approvals found for native address 'jaxl_dxqJ00gEgQazGASFXVRvO4h-Q0_vnaLtuOUoWU'",
  "givenApprovals": [
    {
      "approvalDataItemId": "mwgQ-TfeLoIzyR_C864RKaq2sZmDrx-xBFqcsbDwgDA",
      "approvedAddress": "2cor8AZxm38uc9sVgGaXk4BU2EtB18FqwEq7ivwa7VUa",
      "approvedWincAmount": "8000000000",
      "creationDate": "2024-10-31T14:55:46.889Z",
      "payingAddress": "jaxl_dxqJ00gEgQazGASFXVRvO4h-Q0_vnaLtuOUoWU",
      "usedWincAmount": "0"
    },
    {
      "approvalDataItemId": "pdXRmDue07mX77m7GcgAz2z0esym8H_2-hshs8H454k",
      "approvedAddress": "2cor8AZxm38uc9sVgGaXk4BU2EtB18FqwEq7ivwa7VUa",
      "approvedWincAmount": "8000000000",
      "creationDate": "2024-10-31T15:04:15.898Z",
      "payingAddress": "jaxl_dxqJ00gEgQazGASFXVRvO4h-Q0_vnaLtuOUoWU",
      "usedWincAmount": "0"
    }
  ],
  "receivedApprovals": []
}
```

revoke all active approvals to a given address:

```sh
> node lib/esm/cli/cli.js revoke-credits --address 2cor8AZxm38uc9sVgGaXk4BU2EtB18FqwEq7ivwa7VUa --wallet-file ../wallets/zarweave-keyfile-jaxl_dxqJ00gEgQazGASFXVRvO4h-Q0_vnaLtuOUoWU_dereks.json      
{
  "message": "Revoked credit share approvals!",
  "revokedApprovals": [
    {
      "approvalDataItemId": "mwgQ-TfeLoIzyR_C864RKaq2sZmDrx-xBFqcsbDwgDA",
      "approvedAddress": "2cor8AZxm38uc9sVgGaXk4BU2EtB18FqwEq7ivwa7VUa",
      "approvedWincAmount": "8000000000",
      "creationDate": "2024-10-31T14:55:46.889Z",
      "payingAddress": "jaxl_dxqJ00gEgQazGASFXVRvO4h-Q0_vnaLtuOUoWU",
      "usedWincAmount": "0",
      "inactiveReason": "revoked",
      "inactiveDate": "2024-10-31T15:08:22.317Z",
      "revokeDataItemId": "rGSmUJmYALnMTdGF-V_nnwKLjxgMpqY4neNZByQSa_U"
    },
    {
      "approvalDataItemId": "pdXRmDue07mX77m7GcgAz2z0esym8H_2-hshs8H454k",
      "approvedAddress": "2cor8AZxm38uc9sVgGaXk4BU2EtB18FqwEq7ivwa7VUa",
      "approvedWincAmount": "8000000000",
      "creationDate": "2024-10-31T15:04:15.898Z",
      "payingAddress": "jaxl_dxqJ00gEgQazGASFXVRvO4h-Q0_vnaLtuOUoWU",
      "usedWincAmount": "0",
      "inactiveReason": "revoked",
      "inactiveDate": "2024-10-31T15:08:22.317Z",
      "revokeDataItemId": "rGSmUJmYALnMTdGF-V_nnwKLjxgMpqY4neNZByQSa_U"
    }
  ]
}
```

using payment approvals, by default on upload the CLI will use any received approvals before the signer's balance. the SDK will NOT do so, but integrating apps can choose to do so. To override the paid-by behavior CLI users can supply --paid-by,

```sh
> node lib/esm/cli/cli.js upload-file -f ../uploads/460KB.txt --wallet-file ../wallets/sol.keypair.2cor8AZxm38uc9sVgGaXk4BU2EtB18FqwEq7ivwa7VUa.json --token solana --paid-by jaxl_dxqJ00gEgQazGASFXVRvO4h-Q0_vnaLtuOUoWU
```

or use one of two flags,

--ignore-approvals will skip any check for approvals, using only the signer's balance
--use-signer-balance-first will put the signer at the start of the paid-by list, and then put any received approvals after. this is a helper for using the end of a user's balance (dust)

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -112,6 +112,8 @@ services:
       start_period: 10s
 
   payment-service:
+    # build:
+    #   context: ../payment-service
```
Comment (dtfiedler) [inline docker-compose.yml:116]: can just remove these - shouldn't be building here at all

File: src/common/upload.ts [hunk 2]
```diff
@@ -37,6 +40,11 @@ import {
 import { TurboHTTPService } from './http.js';
 import { TurboWinstonLogger } from './logger.js';
 
+export const createDelegatedPaymentApprovalTagName = 'x-approve-payment';
+export const approvalAmountTagName = 'x-amount';
+export const approvalExpiresBySecondsTagName = 'x-expires-seconds';
+export const revokeDelegatePaymentApprovalTagName = 'x-delete-payment-approval';
```
Comment (dtfiedler) [inline src/common/upload.ts:46]: nit: this could be a map to avoid redundant variable names

**Review by dtfiedler — DISMISSED** [review]

File: src/common/upload.ts (hunk 2 above)
Comment (fedellen) [reply src/common/upload.ts]: [d7a6f5e](https://github.com/ardriveapp/turbo-sdk/pull/196/commits/d7a6f5e8a8aebc717cab30d43c982c04ef422e5b)

File: docker-compose.yml (hunk 1 above)
Comment (fedellen) [reply docker-compose.yml]: [d7a6f5e](https://github.com/ardriveapp/turbo-sdk/pull/196/commits/d7a6f5e8a8aebc717cab30d43c982c04ef422e5b)

**Review by arielmelendez — APPROVED** [review]

**Pull request merged**